### PR TITLE
[airthings] Airthings bluetooth binding initial contribution

### PR DESCRIPTION
**Pull request by paulianttila**
Signed-off-by: Pauli Anttila <pauli.anttila@gmail.com>

### CODE REVIEW

**Comment by paulianttila** [comment]
Waiting fixes #6858, #6857, #6465 and #6861

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    // Channel IDs
+    public static final String CHANNEL_ID_HUMIDITY = "humidity";
+    public static final String CHANNEL_ID_TEMPERATURE = "temperature";
+    public static final String CHANNEL_ID_PRESSURE = "pressure";
+    public static final String CHANNEL_ID_CO2 = "co2";
+    public static final String CHANNEL_ID_TVOC = "tvoc";
+    public static final String CHANNEL_ID_RADON_ST_AVG = "radon_st_avg";
+    public static final String CHANNEL_ID_RADON_LT_AVG = "radon_lt_avg";
+
+    public static final Unit<Dimensionless> PARTS_PER_BILLION = new TransformedUnit<>(SmartHomeUnits.ONE,
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java]: Wouldn't it make more sense to add these to the core UoM bundle?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsHandlerFactory.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            .singleton(AirthingsBindingConstants.THING_TYPE_AIRTHINGS_WAVE_PLUS);
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    @Nullable
+    protected ThingHandler createHandler(Thing thing) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsHandlerFactory.java]:
```suggestion
    protected @Nullable ThingHandler createHandler(Thing thing) {
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusDataParser.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                    co2 = intFromBytes(data[12], data[13]);
+                    tvoc = intFromBytes(data[14], data[15]);
+                } else {
+                    throw new AirthingsParserException(
+                            String.format("Unsupported data structure version '%d'", version));
+                }
+            } else {
+                throw new AirthingsParserException(String.format("Illegal data structure length '%d'", data.length));
+            }
+        } catch (RuntimeException e) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusDataParser.java]: What could happen here that you need to catch `RuntimeException`?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+@NonNullByDefault
+public class AirthingsWavePlusHandler extends BeaconBluetoothHandler {
+
+    private static final String DATA_UUID = "b42e2a68-ade7-11e4-89d3-123b93f75cba";
+
+    private final Logger logger = LoggerFactory.getLogger(AirthingsWavePlusHandler.class);
+    private final UUID uuid = UUID.fromString(DATA_UUID);
+
+    private volatile Boolean servicesResolved = false;
+    private @NonNullByDefault({}) AirthingsConfiguration configuration;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: It has been agreed that using `@NonNullByDefault({})` should not be used in these cases. You could use an `Optional` and `configuration.ifPresent(...);` instead of `@Nullable` and a null-check.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    private @NonNullByDefault({}) ScheduledFuture<?> scheduledTask;
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Please make this `@Nullable` and apply null-checks where necessary.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        logger.debug("Read data from device {}...", address);
+        BluetoothCharacteristic characteristic = device.getCharacteristic(uuid);
+        if (!device.readCharacteristic(characteristic)) {
+            logger.debug("Read data from device {} failed", address);
+            disconnect();
+        }
+    }
+
+    private void printServices() {
+        if (logger.isDebugEnabled()) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: I think you can omit this check (or is `getServices` expensive?).

and

```
device.getServices().forEach(service -> logger.debug("Device {} Service '{}'", address, service));
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+				<description>States how often a refresh shall occur in seconds. This could have impact to battery lifetime</description>
+				<default>300</default>
+			</parameter>
+		</config-description>
+	</thing-type>
+
+	<channel-type id="airthings_humidity">
+		<item-type>Number:Dimensionless</item-type>
+		<label>Humidity</label>
+		<description>Humidity level (%%)</description>
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml]: Applies to all channels: Since you use UoM, adding the unit in the description is not needed (and probably wrong if the user selected a different unit).

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 4 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: ConfigMapper does not support Optionals, should how this should be then handled? Is there an example?

Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]:
```suggestion
    private Optional<AirthingsConfiguration> configuration = Optional.empty();
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    private @NonNullByDefault({}) ScheduledFuture<?> scheduledTask;
+
+    public AirthingsWavePlusHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        configuration = getConfigAs(AirthingsConfiguration.class);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]:
```suggestion
        configuration = Optional.of(getConfigAs(AirthingsConfiguration.class));
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                scheduledTask = null;
+            }
+            servicesResolved = false;
+        }
+    }
+
+    private void startScheduledTask() {
+        if (scheduledTask == null) {
+            logger.debug("Start scheduled task to read device in every {} seconds", configuration.refreshInterval);
+            scheduledTask = scheduler.scheduleWithFixedDelay(this::execute, 10, configuration.refreshInterval,
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]:
```suggestion
            configuration.ifPresent(cfg -> {
                 logger.debug("Start scheduled task to read device in every {} seconds", cfg.refreshInterval);
                 scheduledTask = scheduler.scheduleWithFixedDelay(this::execute, 10, cfg.refreshInterval, TimeUnit.SECONDS);
                });
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+    public AirthingsWavePlusHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        configuration = getConfigAs(AirthingsConfiguration.class);
+        logger.debug("Using configuration: {}", configuration);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]:
```suggestion
        configuration.ifPresent(cfg -> logger.debug("Using configuration: {}", cfg));
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 4 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: This is not complete, I  just wanted to explain the pattern.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java (hunk 1 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java]: PR done

Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java]: PR is merged, but if I have understood correctly, next core version will be 3.0 and as this binding is targetted to be in 2.5.x, I need to keep those in?

Comment (J-N-K) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java]: Yes. Unfortunately that is probably the case.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+            logger.debug("Disconnect from device {} failed", address);
+        }
+    }
+
+    private void discoverServices() {
+        logger.debug("Discover services for device {}", address);
+        if (!device.discoverServices()) {
+            logger.debug("Discovering services failed");
+            disconnect();
+        }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: In the Bluez binding, there is no way initiate a service discovery manually, instead happens automatically when a device is connected to. As such, the `discoverServices` implementation always returns false. So you cant rely on a return of `false` as meaning something failed.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+ */
+@NonNullByDefault
+public class AirthingsWavePlusHandler extends BeaconBluetoothHandler {
+
+    private static final String DATA_UUID = "b42e2a68-ade7-11e4-89d3-123b93f75cba";
+
+    private final Logger logger = LoggerFactory.getLogger(AirthingsWavePlusHandler.class);
+    private final UUID uuid = UUID.fromString(DATA_UUID);
+
+    private volatile Boolean servicesResolved = false;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Why not use primitive boolean?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+                updateStatus(ThingStatus.ONLINE);
+                try {
+                    updateChannels(new AirthingsWavePlusDataParser(characteristic.getValue()));
+                } catch (AirthingsParserException e) {
+                    logger.warn("Data parsing error occured, when parsing data from device {}, cause {}", address,
+                            e.getMessage(), e);
+                }
+            } else {
+                logger.debug("Characteristic {} from device {} failed", characteristic.getUuid(), address);
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No response from device");
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Wouldn't the Thing status almost immediately switch back to ONLINE when BeaconBluetoothHandler receives the next scanned rssi value?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+	</channel-type>
+	<channel-type id="airthings_pressure">
+		<item-type>Number:Pressure</item-type>
+		<label>Pressure</label>
+		<description>Pressure</description>
+		<state readOnly="true" pattern="%.0f %unit%" />
+	</channel-type>
+	<channel-type id="airthings_co2">
+		<item-type>Number:Dimensionless</item-type>
+		<label>CO2 Level</label>
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/resources/ESH-INF/thing/airthings.xml]: Would using unicode subscripts work here?
```suggestion
		<label>CO₂ Level</label>
```

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    public void dispose() {
+        try {
+            super.dispose();
+        } finally {
+            if (scheduledTask != null) {
+                scheduledTask.cancel(true);
+                scheduledTask = null;
+            }
+            servicesResolved = false;
+        }
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: I would suggest calling `super.dispose()` last to give tasks that you might be running at the moment a cleaner shutdown. Disposing the super first is like pulling the rug out from under any asynchronous tasks your handler might be running.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    }
+
+    private void startScheduledTask() {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: why isn't this called during initialization?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 15 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Slipped through my attention. I agree.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 12 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: It's used also for synchronisation, which primitive type doesn't support.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 13 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Sure, if scan result is received.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 16 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: If device is not yet discovered (which is not happened during initialisation), first task execution will "fail". If user have configured poll intervall e.g. for an hour, next/first poll will happen rather late.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 11 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Will Bluez send event when services are discovered (onServicesDiscovered)?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 16 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: You only start the scheduled task when the handler receives a connections state change to the `DISCOVERED` state. Which requires two things to happen:
1. That the device must switch to the DISCOVERED state, which only happens when the device is initially seen by an adapter. Also, almost all device immediately switch from DISCOVERED to DISCONNECTED, so simply checking if the device is in the DISCOVERED state will often not work.
2. That the handler must be listening to the device when it changes to its DISCOVERED state. This requires that the handler be initialized and running before the device is discovered by a bluetooth adapter. Which means that the handler will not work if the user tries to create the handler from an inbox discovery; If it is in the inbox, that means it was already discovered!

Not to mention that there is no guarantee on openhab startup that the `Airthings` handler will be started before the bluetooth device handlers. In fact, by virtue of the bluetooth handlers using the bluetooth adapters as a bridge, it is almost guaranteed that the adapters will start first. If the device is in range during that time your handler could very well miss the DISCOVERED event.

To summarize: make sure that your scheduled task lifecycle is tied to the handler and not dependent on device state.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 11 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Yep, in Bluez `onServicesDiscovered` will always get called after a successful connection (assuming the services get discovered properly)

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+    private void execute() {
+        ConnectionState connectionState = device.getConnectionState();
+        logger.debug("Device {} state is {}", address, connectionState);
+        switch (connectionState) {
+            case DISCONNECTED:
+                connect();
+                break;
+            case CONNECTED:
+                synchronized (servicesResolved) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: You REALLY shouldn't be synchronizing on a Boolean instance.
This [post](https://stackoverflow.com/questions/10324272/why-is-it-not-a-good-practice-to-synchronize-on-boolean) give an elegant description of why.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+@NonNullByDefault
+public class AirthingsWavePlusHandler extends BeaconBluetoothHandler {
+
+    private static final String DATA_UUID = "b42e2a68-ade7-11e4-89d3-123b93f75cba";
+    private static final int CHECK_PERIOD_SEC = 10;
+
+    private final Logger logger = LoggerFactory.getLogger(AirthingsWavePlusHandler.class);
+    private final UUID uuid = UUID.fromString(DATA_UUID);
+
+    private AtomicInteger sinceLastReadSec = new AtomicInteger();
```
Comment (cpmeister) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: I don't know what purpose this is serving that executing periodic already doesn't. Couldn't this be replaced with a boolean flag?
Set it to true at the start of the scheduled task.
Set it to false when value data is successfully read.
But even that boolean flag could be replaced as states in an overall state machine.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+                            disconnect();
+                        }
+                        break;
+                    default:
+                        break;
+                }
+            default:
+                break;
+        }
+    }
```
Comment (paulianttila) [inline bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: > You've made a rather complex state machine and I think it should be simplified.

I wouldn't say that this is complex.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+            new RationalConverter(BigInteger.ONE, BigInteger.valueOf(1000000000)));
```
[the remark this answers is not in the document]
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java]: These are all ready merged to the OH core, so I like to use same. These will be removed from the binding when 3.0 is published.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 19 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Maybe complex is the wrong word. Let's just say the code can't be understood at a glance.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 18 above)
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: State machine is updated every 10 second by the scheduler. If polling period is e.g. set to one hour by the user, first read (or second actually) is done after hour and so on. But if poll fails, retry is done after 10 second not after one hour. This functionality can't be implemented by boolean flag or state.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsBindingConstants.java]: understood

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+    private enum ServiceState {
+        NOT_RESOLVED,
+        RESOLVING,
+        RESOLVED,
+    }
+
+    private enum ReadState {
+        IDLE,
+        READING,
+    }
```
[the remark this answers is not in the document]
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Most probably yes, but I think that service resolving shouldn't be even duty of the bindings but handled automatically by the BluetoothDevice after connect. Now, every binding need to handle it, which doesn't make sense to me. As you are now simplifying and refactoring BluetoothDevice, I propose that you consider that and we could remove whole service resolving staff out from the binding(s).

[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: I agree and am completely on board with that. One potential issue that would need to be eventually addressed though are bluetooth devices that might not support the GATT protocol at all.

So do you want this binding to be merged before or after the refactoring PR?

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 18 above)
Comment (cpmeister) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: Guess you are right. It didn't occur to me that the scheduler period and read period were different.

File: bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (paulianttila) [reply bundles/org.openhab.binding.bluetooth.airthings/src/main/java/org/openhab/binding/bluetooth/airthings/internal/AirthingsWavePlusHandler.java]: My wish is that binding will be merged before refactoring PR.
